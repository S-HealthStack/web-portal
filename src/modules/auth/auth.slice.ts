import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { push } from 'connected-react-router';
import jwtDecode from 'jwt-decode';
import API, { SigninResponse } from 'src/modules/api';
import applyDefaultApiErrorHandlers from 'src/modules/api/applyDefaultApiErrorHandlers';
import { Path } from 'src/modules/navigation/store';
import { showSnackbar } from 'src/modules/snackbar/snackbar.slice';
import { AppThunk, RootState } from 'src/modules/store';
import {
  mockStudyIds,
  selectedStudyIdSelector,
  studiesSlice,
} from 'src/modules/studies/studies.slice';
import {
  getRoleForStudy,
  rolesListFromApi,
  roleToApi,
  isGlobalRoleType,
  allowedRoleTypes,
} from './userRole';

export const STORAGE_TOKEN_KEY = 'auth_token';

const isValidEmail = (v: string) => v.includes('samsung');

type AuthTokenPayload = {
  email: string;
  roles: string[];
};

export const decodeAuthToken = (jwt: string): AuthTokenPayload => jwtDecode<AuthTokenPayload>(jwt);

API.mock.provideEndpoints({
  signin({ email }) {
    if (isValidEmail(email)) {
      let roles = ['team-admin'];

      const roleFromEmail = allowedRoleTypes.find((r) => email.includes(r));
      if (roleFromEmail) {
        roles = isGlobalRoleType(roleFromEmail)
          ? [roleFromEmail]
          : mockStudyIds.map((projectId) => roleToApi({ role: roleFromEmail, projectId }));
      }

      const header = window.btoa(JSON.stringify({}));
      const payload = window.btoa(
        JSON.stringify({
          email,
          roles,
        })
      );
      const jwt = `${header}.${payload}`;
      return API.mock.response<SigninResponse>({
        id: email,
        jwt,
        email,
        roles,
      });
    }

    return API.mock.failedResponse({
      status: 401,
    });
  },
  resetPassword(body) {
    if (isValidEmail(body.email) && body.resetToken) {
      return API.mock.response(undefined);
    }

    return API.mock.failedResponse({
      status: 401,
    });
  },
});

interface AuthState {
  authToken?: string;
}

export const getStateFromAuthToken = (authToken: string) => {
  const { roles, email } = decodeAuthToken(authToken);
  return {
    authToken,
    userRoles: rolesListFromApi(roles),
    username: email,
  };
};

export const loadInitialStateFromStorage = (): AuthState => {
  const authToken = localStorage.getItem(STORAGE_TOKEN_KEY) || undefined;

  return { authToken };
};

const initialState: AuthState = {
  ...loadInitialStateFromStorage(),
};

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    authSuccess(_, { payload }: PayloadAction<{ authToken: string }>) {
      return getStateFromAuthToken(payload.authToken);
    },
    clearAuth(state) {
      state.authToken = undefined;
    },
  },
});

const { authSuccess, clearAuth } = authSlice.actions;

export const authTokenPayloadSelector = createSelector(
  (state: RootState) => state.auth.authToken,
  (authToken) => {
    if (!authToken) {
      return {};
    }

    try {
      const { email, roles } = decodeAuthToken(authToken);
      return {
        email,
        roles: rolesListFromApi(roles),
      };
    } catch (err) {
      console.error(`Failed to decode jwt ${err}`);

      // TODO: better way to sign out?
      // we can use dispatch from store directly, but for now it's fine as is
      localStorage.removeItem(STORAGE_TOKEN_KEY);
      window.location.reload();

      return {};
    }
  }
);

export const signout = (): AppThunk => (dispatch) => {
  localStorage.removeItem(STORAGE_TOKEN_KEY);
  dispatch(clearAuth());
  dispatch(studiesSlice.actions.reset());
};

export const signin =
  ({
    email,
    password,
    rememberUser,
  }: {
    email: string;
    password: string;
    rememberUser?: boolean;
  }): AppThunk<Promise<void>> =>
  async (dispatch, getState) => {
    const res = await API.signin({ email, password });
    const { jwt: authToken } = res.data;

    const isFirstTime = false; // TODO: need to have API or make it defined somehow

    if (rememberUser) {
      localStorage.setItem(STORAGE_TOKEN_KEY, authToken);
    } else {
      localStorage.removeItem(STORAGE_TOKEN_KEY);
    }

    dispatch(authSuccess({ authToken }));

    const tokenPayload = authTokenPayloadSelector(getState());

    if (tokenPayload.roles?.some((r) => r.role === 'team-admin') && isFirstTime) {
      dispatch(push(Path.CreateStudy));
    } else {
      dispatch(push(Path.Root));
    }
  };

export const activateAccount =
  ({
    email,
    name,
    password,
    resetToken,
  }: {
    email: string;
    name: string;
    password: string;
    resetToken: string;
  }): AppThunk<Promise<void>> =>
  async (dispatch) => {
    try {
      const res = await API.resetPassword({
        email,
        password,
        resetToken,
        profile: {
          name,
          status: 'active',
        },
      });
      res.checkError();

      await dispatch(signin({ email, password, rememberUser: true }));
    } catch (err) {
      if (!applyDefaultApiErrorHandlers(err, dispatch)) {
        dispatch(showSnackbar({ text: 'Failed to activate account' }));
      }
    }
  };

export const isAuthorizedSelector = (state: RootState) => !!state.auth.authToken;

export const userNameSelector = createSelector(
  authTokenPayloadSelector,
  (pl) => pl?.email?.split('@')[0] || pl?.email
);

export const userRoleSelector = createSelector(
  [authTokenPayloadSelector, selectedStudyIdSelector],
  (pl, selectedStudyId) => getRoleForStudy(pl.roles || [], selectedStudyId)
);

export default authSlice.reducer;
