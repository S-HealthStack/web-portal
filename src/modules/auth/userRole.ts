export type UserRole = {
  role: 'team-admin' | 'researcher' | 'project-owner';
  projectId?: string;
};

export type RoleType = UserRole['role'];

export const allowedRoleTypes: Readonly<RoleType[]> = [
  'team-admin',
  'researcher',
  'project-owner',
] as const;

export const isValidRoleType = (rt: string): rt is RoleType =>
  allowedRoleTypes.includes(rt as RoleType);

const globalRoleTypes: Readonly<RoleType[]> = ['team-admin'] as const;

export const isGlobalRoleType = (role: RoleType) => globalRoleTypes.includes(role);

export const roleFromApi = (roleStr: string) => {
  const [projectId, role] = (roleStr.includes(':') ? roleStr : `:${roleStr}`).split(':');
  if (!isValidRoleType(role)) {
    console.warn(`Invalid role ${roleStr}`);
    return undefined;
  }
  if (!isGlobalRoleType(role as RoleType) && !projectId) {
    console.warn(`Project required for role ${roleStr}`);
    return undefined;
  }

  return {
    role,
    projectId: projectId || undefined,
  } as UserRole;
};

export const rolesListFromApi = (roles: string[]): UserRole[] =>
  roles.map(roleFromApi).filter(Boolean) as UserRole[];

export const roleToApi = (role: UserRole) =>
  isGlobalRoleType(role.role) ? role.role : `${role.projectId}:${role.role}`;

export const getRoleForStudy = (
  roles: UserRole[],
  studyId: string | undefined
): UserRole | undefined =>
  roles.find((r) => isGlobalRoleType(r.role)) || roles.find((r) => r.projectId === studyId);

export const getRoleLabel = (role: UserRole) =>
  ({ 'team-admin': 'Team Admin', researcher: 'Researcher', 'project-owner': 'Project Owner' }[
    role.role
  ]);

export const getRoleFunction = (role: RoleType) =>
  role === 'researcher' ? 'principal_investigator' : 'study_operator';
