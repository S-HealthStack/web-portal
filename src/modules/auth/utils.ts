import jwtDecode from 'jwt-decode';

export const STORAGE_TOKEN_KEY = 'auth_token';
export const STORAGE_REFRESH_TOKEN_KEY = 'refresh_token';

type AuthTokenPayload = {
  email: string;
  roles: string[];
};

export const decodeAuthToken = (jwt: string): AuthTokenPayload => jwtDecode<AuthTokenPayload>(jwt);

export const STORAGE_ALREADY_BEEN_AUTH_KEY = 'already_been_auth';
