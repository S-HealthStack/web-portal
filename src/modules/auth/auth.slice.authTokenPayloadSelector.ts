import { createSelector } from '@reduxjs/toolkit';

import { RootState } from 'src/modules/store';

import { userRolesListFromApi } from 'src/modules/auth/userRole';
import { decodeAuthToken } from 'src/modules/auth/utils';
import { signout } from 'src/modules/auth/auth.slice.signout';

export const getUserDataFromAuthToken = (authToken: string | undefined) => {
  if (!authToken) {
    return {};
  }

  try {
    const { email, roles } = decodeAuthToken(authToken);
    return {
      email,
      roles: userRolesListFromApi(roles),
    };
  } catch (err) {
    console.error(`Failed to decode jwt ${err}`);

    signout();

    return {};
  }
};

export const authTokenPayloadSelector = createSelector(
  (state: RootState) => state.auth.authToken,
  (authToken) => getUserDataFromAuthToken(authToken)
);
