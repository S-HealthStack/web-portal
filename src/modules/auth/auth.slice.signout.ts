import { clearAuth } from 'src/modules/auth/auth.slice';
import { reset } from 'src/modules/studies/studies.slice';
import { AppThunk } from 'src/modules/store';
import { STORAGE_TOKEN_KEY, STORAGE_REFRESH_TOKEN_KEY } from 'src/modules/auth/utils';

export const signout = (): AppThunk => (dispatch) => {
  localStorage.removeItem(STORAGE_TOKEN_KEY);
  localStorage.removeItem(STORAGE_REFRESH_TOKEN_KEY);
  dispatch(clearAuth());
  dispatch(reset());
};
