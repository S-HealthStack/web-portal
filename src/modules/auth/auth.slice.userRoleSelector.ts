import { createSelector } from '@reduxjs/toolkit';

import { authTokenPayloadSelector } from 'src/modules/auth/auth.slice.authTokenPayloadSelector';
import { getRolesForStudy } from 'src/modules/auth/userRole';
import { selectedStudyIdSelector } from 'src/modules/studies/studies.slice';

export const userRoleSelector = createSelector(
  [authTokenPayloadSelector, selectedStudyIdSelector],
  (pl, selectedStudyId) => getRolesForStudy(pl?.roles || [], selectedStudyId)
);
