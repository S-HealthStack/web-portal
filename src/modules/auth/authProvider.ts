import { AppDispatch } from 'src/modules/store';
import { store } from 'src/modules/store/store';
import API from 'src/modules/api';

import { signout } from './auth.slice';

API.setAuthProvider({
  getBearerToken() {
    return store.getState().auth.authToken;
  },
  onUnauthorizedError() {
    (store.dispatch as AppDispatch)(signout());
  },
});
