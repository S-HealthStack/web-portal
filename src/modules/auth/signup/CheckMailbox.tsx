import React, { useCallback, useMemo } from 'react';
import styled from 'styled-components';

import { colors, px, typography } from 'src/styles';
import EmailSuccessImg from 'src/assets/illustrations/email_success.svg';
import Link from 'src/common/components/Link';
import Card from 'src/common/components/Card';
import ResultMessage from 'src/modules/auth/signup/ResultMessage';
import { SnackbarContainer } from 'src/modules/snackbar';
import { useAppDispatch } from 'src/modules/store';
import { resendVerificationEmail } from 'src/modules/auth/auth.slice';

const Container = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100%;
  width: 100%;
`;

const ContentCard = styled(Card)`
  width: 90%;
  max-width: ${px(800)};
  min-height: ${px(546)};
  > div {
    display: flex;
    align-items: center;
    justify-content: center;
  }
`;

const ResendEmailOffer = styled.div`
  ${typography.bodySmallRegular};
  margin-top: ${px(53)};

  a {
    ${typography.bodySmallSemibold};
    color: ${colors.textPrimaryBlue};
    display: inline-block;
    margin-left: ${px(8)};
  }
`;

const CheckMailbox = () => {
  const email = useMemo(() => new URLSearchParams(window.location.search).get('email') || '', []);
  const dispatch = useAppDispatch();

  const handleResend = useCallback(
    (evt: React.MouseEvent<HTMLAnchorElement>) => {
      evt.preventDefault();
      evt.stopPropagation();

      dispatch(resendVerificationEmail({ email }));
    },
    [dispatch, email]
  );

  return (
    <>
      <Container>
        <ContentCard>
          <ResultMessage
            picture={<EmailSuccessImg />}
            title="Check your mailbox"
            description={
              <>
                We’ve emailed an account activation link to
                <br />
                <strong>{email}</strong>
              </>
            }
          >
            <ResendEmailOffer>
              Didn’t get the email?
              <Link data-testid="signup-check-mailbox-resend" to="/" onClick={handleResend}>
                Resend the email
              </Link>
            </ResendEmailOffer>
          </ResultMessage>
        </ContentCard>
      </Container>
      <SnackbarContainer useSimpleGrid />
    </>
  );
};

export default CheckMailbox;
