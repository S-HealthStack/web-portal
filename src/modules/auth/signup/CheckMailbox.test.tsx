import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import userEvent from '@testing-library/user-event';
import 'jest-styled-components';

import theme from 'src/styles/theme';
import { makeStore } from 'src/modules/store/store';
import { makeHistory } from 'src/modules/navigation/store';
import CheckMailbox from 'src/modules/auth/signup/CheckMailbox';
import { currentSnackbarSelector } from 'src/modules/snackbar/snackbar.slice';
import { SUCCESS_CONFIRMATION_MESSAGE } from 'src/modules/auth/auth.slice';
import { enableFetchMocks } from 'jest-fetch-mock';

beforeAll(() => {
  enableFetchMocks();
});

describe('CheckMailbox', () => {
  let store: ReturnType<typeof makeStore>;
  let history: ReturnType<typeof makeHistory>;

  beforeEach(() => {
    history = makeHistory();
    store = makeStore(history);
    localStorage.setItem('API_URL', 'https://samsung.com/');
  });

  it('should render', async () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <ConnectedRouter history={history}>
            <CheckMailbox />
          </ConnectedRouter>
        </Provider>
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const resend = await screen.findByTestId('signup-check-mailbox-resend');
    expect(resend).not.toBeNull();
    await userEvent.click(resend);

    await waitFor(() =>
      expect(currentSnackbarSelector(store.getState())).toMatchObject({
        text: SUCCESS_CONFIRMATION_MESSAGE,
      })
    );
  });
});
