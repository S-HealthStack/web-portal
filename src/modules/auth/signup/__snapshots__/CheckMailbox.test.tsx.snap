// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckMailbox should render 1`] = `
.c9 {
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: default;
  -webkit-user-drag: none;
}

.c9:hover {
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #FFFFFF;
  box-shadow: 0 0 0.125rem 0rem rgba(0,0,0,0.15);
  border-radius: 0.25rem;
  box-sizing: border-box;
  padding: 1.5rem 1.5rem 1rem;
  overflow: visible;
  position: relative;
}

.c1 > * {
  min-width: 0;
  min-height: 0;
  overflow: visible;
}

.c3 {
  -webkit-transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  opacity: 1;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: -0.875rem;
}

.c4 > * {
  text-align: center;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 15rem;
  height: 15rem;
}

.c6 {
  font-family: Inter,sans-serif;
  font-size: 2rem;
  line-height: 2.625rem;
  font-weight: 600;
  font-style: normal;
  margin: 0;
  margin-top: 2rem;
}

.c7 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 400;
  font-style: normal;
  max-width: 37.5rem;
  margin: 0;
  margin-top: 0.5rem;
}

.c7 strong {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 600;
  font-style: normal;
}

.c10 {
  position: relative;
  height: 0;
  width: 100%;
  pointer-events: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 100%;
  width: 100%;
}

.c2 {
  width: 90%;
  max-width: 50rem;
  min-height: 34.125rem;
}

.c2 > div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c8 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  margin-top: 3.3125rem;
}

.c8 a {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 600;
  font-style: normal;
  color: #4475E3;
  display: inline-block;
  margin-left: 0.5rem;
}

@media (min-width:64rem) {

}

@media (min-width:90rem) {

}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="c1 c2"
      >
        <div
          class="c3"
          data-testid="card-content"
        >
          <div
            class="c4"
          >
            <div
              class="c5"
            >
              <svg />
            </div>
            <h2
              class="c6"
            >
              Check your mailbox
            </h2>
            <p
              class="c7"
            >
              We’ve emailed an account activation link to
              <br />
              <strong />
            </p>
            <div
              class=""
            >
              <div
                class="c8"
              >
                Didn’t get the email?
                <a
                  class="c9"
                  data-testid="signup-check-mailbox-resend"
                  href="/"
                >
                  Resend the email
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c10"
    />
  </div>
</body>
`;
